### PR TITLE
917 incorrect conversion to dataframe for PIVOT statement and PIVOT subqueries

### LINKED ISSUE #917: [DuckDB] incorrect conversion to Pandas/Polars dataframe for PIVOT statement results
### What happens?

I was following the [DuckDB documentation examples for the PIVOT statement](https://duckdb.org/docs/sql/statements/pivot.html), and discovered a bug when converting the results of a PIVOT statement to a Pandas dataframe

### To Reproduce

Notebook setup:
```python
import duckdb
conn = duckdb.connect()

%load_ext sql
%sql conn --alias duckdb-mem
```

Table defined as per DuckDB documentation example:
```
%%sql
CREATE OR REPLACE TABLE Cities(Country VARCHAR, Name VARCHAR, Year INT, Population INT);
INSERT INTO Cities VALUES ('NL', 'Amsterdam', 2000, 1005);
INSERT INTO Cities VALUES ('NL', 'Amsterdam', 2010, 1065);
INSERT INTO Cities VALUES ('NL', 'Amsterdam', 2020, 1158);
INSERT INTO Cities VALUES ('US', 'Seattle', 2000, 564);
INSERT INTO Cities VALUES ('US', 'Seattle', 2010, 608);
INSERT INTO Cities VALUES ('US', 'Seattle', 2020, 738);
INSERT INTO Cities VALUES ('US', 'New York City', 2000, 8015);
INSERT INTO Cities VALUES ('US', 'New York City', 2010, 8175);
INSERT INTO Cities VALUES ('US', 'New York City', 2020, 8772);
```

If we run the following PIVOT statement using the the DuckDB Python client directly and convert to a Pandas dataframe, we get the expected three rows seen in the DuckDB docs:

```python
conn.sql("PIVOT Cities ON Year USING SUM(Population)").df()
```
native_duckdb_df

When we run the same query with JupySQL, getting the results back as a ResultSet, we get the expected 3 rows.
```
%%sql
PIVOT Cities ON Year USING SUM(Population)
```

But when we force this to be converted into a Pandas dataframe, we only get one row back:
```
%config SqlMagic.autopandas = True
```

```
%%sql
PIVOT Cities ON Year USING SUM(Population)
```

jupysql_df

The same behaviour happens when converting to Polars too.

### OS:

Ubuntu 23.04

### JupySQL Version:

0.10.2

### Full Name:

Ned Letcher

### Affiliation:

Thoughtworks

### PR DESCRIPTION
**Pull request by bryannho**
## Describe your changes
- Modified `_statement_is_select()` in `resultset.py` to identify 'PIVOT' as a 'SELECT' statement (with native duckdb connection, 'SELECT' statements must be executed twice in order to receive the correct results as a dataframe)
- Modified `_convert_to_dataframe()` in `resultset.py` to begin a new transaction before executing any "PIVOT" queries or subqueries. Every 'SELECT' or similar statement is run twice, but for 'PIVOT' specifically, duckDB thinks a new transaction is occurring inside an existing one. So we need to start a fresh transaction in order to avoid this error.
- Added tests to ensure 'PIVOT' queries with dataframe conversion retrieve the correct results
- Added tests to ensure nested queries including 'PIVOT' do not trigger `InvalidInputException` or `TransactionContext Error`

## Issue number

Closes #917 

## Checklist before requesting a review

- [x] Performed a self-review of my code
- [x] Formatted my code with [`pkgmt format`](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#linting-formatting)
- [x] Added [tests](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#testing) (when necessary).
- [x] Added [docstring](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#documenting-changes-and-new-features) documentation and update the [changelog](https://ploomber-contributing.readthedocs.io/en/latest/contributing/pr.html#changelog) (when needed)

----
:books: Documentation preview :books:: https://jupysql--921.org.readthedocs.build/en/921/

### CODE REVIEW

**Comment by bryannho** [comment]
@edublancas could you take a look at the CI? this PR is only failing for windows and it's an error i've never seen before. will continue troubleshooting. is this a known error?

**Comment by edublancas** [comment]
this is an intermittent error:

```
 FAILED src/tests/test_magic_plot.py::test_hist_bin[png] - _tkinter.TclError: Can't find a usable init.tcl in the following directories: 
    {C:\hostedtoolcache\windows\Python\3.11.6\x64\tcl\tcl8.6}

C:/hostedtoolcache/windows/Python/3.11.6/x64/tcl/tcl8.6/init.tcl: couldn't read file "C:/hostedtoolcache/windows/Python/3.11.6/x64/tcl/tcl8.6/init.tcl": No error
couldn't read file "C:/hostedtoolcache/windows/Python/3.11.6/x64/tcl/tcl8.6/init.tcl": No error
    while executing
"source C:/hostedtoolcache/windows/Python/3.11.6/x64/tcl/tcl8.6/init.tcl"
    ("uplevel" body line 1)
    invoked from within
"uplevel #0 [list source $tclfile]"

```

you can push an empty commit `git commit -m ci --allow-empty` to re-trigger the CI. or retry the jobs that failed

File: src/tests/test_resultset.py [hunk 1]
```diff
@@ -692,3 +692,116 @@ def test_calling_legacy_plotting_functions_displays_warning(
...
+    PIVOT Cities
+    ON Year
+    USING FIRST(Population)
+)
+WHERE "2000" > 1000
+    """
+    )
+
+    error = "InvalidInputException: Invalid Input Error: "
+    "Attempting to execute an unsuccessful or closed pending query result "
```
Comment (neelasha23) [inline src/tests/test_resultset.py:803]: No need of testing error not in result. just add assert for the expected results

**Comment by neelasha23** [comment]
Added minor comment.
Tested locally and it works!
@edublancas @bryannho

File: src/tests/test_resultset.py (hunk 1 above)
Comment (bryannho) [reply src/tests/test_resultset.py]: Removed this test!

**Comment by bryannho** [comment]
Fixed according to @neelasha23's comment! Ready for review @neelasha23 @nooodle-soup @edublancas

**Comment by edublancas** [comment]
please review @nooodle-soup @neelasha23

**Review by neelasha23 — DISMISSED** [review]

**Review by edublancas — CHANGES_REQUESTED** [review]
qq: the reported error is

```
Error: TransactionContext Error: cannot start a transaction within a transaction
```

this implies that we're trying to start a transaction while already in one. where in the code are we starting a transaction? seems odd to me that the solution to this problem is to start *another* transaction?

perhaps this is the only solution but I'm wondering if there's a better way to fix it

**Comment by bryannho** [comment]
> this implies that we're trying to start a transaction while already in one. where in the code are we starting a transaction? seems odd to me that the solution to this problem is to start _another_ transaction?

@edublancas I'm not fully sure where in the code we start a transaction. My theory is that DuckDB treats `pivot` queries (but not `select`) as transactions. When we run the query for the first time it is successful, but when converting the results to a DataFrame the query must be re-executed. For some reason the previous transaction isn't ended before the query is re-executed, causing the error. I could be interpreting the code/error wrong so I will have to dig a bit to better understand what's happening.

> perhaps this is the only solution but I'm wondering if there's a better way to fix it

Starting another transaction seems to work because it ends the previous transaction before re-executing the query. I agree this isn't ideal, so I'll look again into

1. Why DuckDB thinks we're beginning a transaction in the first place
2. Another way to work around this issue without starting any new transactions. Perhaps we can just end the current transaction, or if necessary, execute the queries within separate [commits](https://duckdb.org/docs/api/python/reference/#duckdb.DuckDBPyConnection) in the same transaction

**Comment by edublancas** [comment]
@bryannho ok, please spend some time understanding this. but if you don't encounter any better solution, we can merge it as is

**Comment by bryannho** [comment]
@edublancas I did some digging into [DuckDB's docs](https://duckdb.org/duckdb-docs.pdf) and found a few answers:

1. When queries are executed with `execute()`, each is a separate _transaction_. The transaction begins, query is executed, the results are available to through current `connection`, and once the results are fetched/returned, the transaction is closed. Usually results are automatically output to the notebook, but you can also manually fetch them into a Python list using `fetchone()` or `fetchall()`. Once all results are _fetched_, the current transaction is closed. DuckDB explains:

Screen Shot 2023-11-02 at 10 06 08 AM

2. DuckDB processes `PIVOT` statements by converting them into multiple nested `SELECT` statements internally. For this reason, I believe it doesn't automatically close the transaction when you execute the query and output the results. See here:

Screen Shot 2023-11-02 at 10 05 11 AM

Something about the way it re-writes queries is allowing the transaction to stay open after it is executed. So now we know we simply need to close the existing transaction before re-executing another query. We can do this by using `fetchall()`, which [I implemented here](https://github.com/ploomber/jupysql/pull/921/files#diff-ee061806ed7a4f5a79d8571f64241031f5c61240dbe59e548de0666a7996ab88R509-R514) to close the transaction. This behavior is preferred to starting a new transaction with `begin()`.

**Review by edublancas — APPROVED** [review]

**Comment by edublancas** [comment]
@bryannho thanks a lot for investigating this. this looks like a much better solution!

**Pull request merged**